### ISSUE TITLE
fails to build with missing file

### ISSUE THREAD

**Issue by wjwwood** [ament/uncrustify#1]
When building on my Trusty VM with `./src/ament/ament_tools/scripts/ament.py build ./src --build-tests`:

```
Process package 'uncrustify' with context:
--------------------------------------------------------------------------------
 source_space => /home/william/ros2_ws/src/ament/uncrustify
  build_space => /home/william/ros2_ws/build/uncrustify
install_space => /home/william/ros2_ws/install
   make_flags => -j4, -l4
  build_tests => True
--------------------------------------------------------------------------------
+++ Building 'uncrustify'
Running cmake because arguments have changed.
==> '. /home/william/ros2_ws/build/uncrustify/ament_cmake__build.sh && /usr/bin/cmake /home/william/ros2_ws/src/ament/uncrustify -DCMAKE_INSTALL_PREFIX=/home/william/ros2_ws/install -DAMENT_ENABLE_TESTING=1' in '/home/william/ros2_ws/build/uncrustify'
-- The C compiler identification is GNU 4.8.2
-- The CXX compiler identification is GNU 4.8.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    AMENT_ENABLE_TESTING


-- Build files have been written to: /home/william/ros2_ws/build/uncrustify
==> '. /home/william/ros2_ws/build/uncrustify/ament_cmake__build.sh && /usr/bin/make -j4 -l4' in '/home/william/ros2_ws/build/uncrustify'
Scanning dependencies of target external_uncrustify
[ 12%] Creating directories for 'external_uncrustify'
[ 25%] No download step for 'external_uncrustify'
[ 50%] [ 50%] No patch step for 'external_uncrustify'
No update step for 'external_uncrustify'
[ 62%] Performing configure step for 'external_uncrustify'
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking for style of include used by make... GNU
checking dependency style of g++... none
checking for gcc... gcc
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... none
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for inttypes.h... (cached) yes
checking for stdint.h... (cached) yes
checking utime.h usability... yes
checking utime.h presence... yes
checking for utime.h... yes
checking for stdbool.h that conforms to C99... yes
checking for _Bool... yes
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for working memcmp... yes
checking for memset... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strdup... yes
checking for strerror... yes
checking for strtol... yes
checking for strtoul... yes
checking whether our compiler supports __func__... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating man/Makefile
config.status: creating man/uncrustify.1
config.status: creating src/uncrustify_version.h
config.status: creating src/config.h
config.status: executing depfiles commands
[ 75%] No build step for 'external_uncrustify'
[ 87%] Performing install step for 'external_uncrustify'
Making install in src
Rebuilding token_names.h
sh: 0: Can't open ../make_token_names.sh
make[4]: *** [token_names.h] Error 127
make[3]: *** [install-recursive] Error 1
make[2]: *** [external_uncrustify-prefix/src/external_uncrustify-stamp/external_uncrustify-install] Error 2
make[1]: *** [CMakeFiles/external_uncrustify.dir/all] Error 2
make: *** [all] Error 2
```

I believe I'm using the latest:

```
$ cd src/ament/uncrustify && git log -1; cd -
commit b9cd1dc9b73eab61904491bf2a565622b34e5cc4
Author: Dirk Thomas <dthomas@osrfoundation.org>
Date:   Tue Apr 14 15:34:49 2015 -0700

    add package.xml and CMakeLists.txt
```

And I've completely cleared my `build` and `install` spaces before.

**Comment by wjwwood** [comment]
This is the output of `tree` for my `build/uncrustify` folder:

```
% tree build/uncrustify
build/uncrustify
├── ament_cmake_args.cache
├── ament_cmake__build.sh
├── CMakeCache.txt
├── CMakeFiles
│   ├── 2.8.12.2
│   │   ├── CMakeCCompiler.cmake
│   │   ├── CMakeCXXCompiler.cmake
│   │   ├── CMakeDetermineCompilerABI_C.bin
│   │   ├── CMakeDetermineCompilerABI_CXX.bin
│   │   ├── CMakeSystem.cmake
│   │   ├── CompilerIdC
│   │   │   ├── a.out
│   │   │   └── CMakeCCompilerId.c
│   │   └── CompilerIdCXX
│   │       ├── a.out
│   │       └── CMakeCXXCompilerId.cpp
│   ├── cmake.check_cache
│   ├── CMakeDirectoryInformation.cmake
│   ├── CMakeOutput.log
│   ├── CMakeRuleHashes.txt
│   ├── CMakeTmp
│   ├── external_uncrustify.dir
│   │   ├── build.make
│   │   ├── cmake_clean.cmake
│   │   ├── DependInfo.cmake
│   │   ├── depend.internal
│   │   ├── depend.make
│   │   └── progress.make
│   ├── Makefile2
│   ├── Makefile.cmake
│   ├── Progress
│   │   ├── 2
│   │   ├── 3
│   │   ├── 4
│   │   ├── 5
│   │   ├── 6
│   │   ├── 7
│   │   ├── 8
│   │   └── count.txt
│   ├── progress.marks
│   └── TargetDirectories.txt
├── cmake_install.cmake
├── external_uncrustify-prefix
│   ├── src
│   │   ├── external_uncrustify-build
│   │   │   ├── config.log
│   │   │   ├── config.status
│   │   │   ├── Makefile
│   │   │   ├── man
│   │   │   │   ├── Makefile
│   │   │   │   └── uncrustify.1
│   │   │   └── src
│   │   │       ├── config.h
│   │   │       ├── Makefile
│   │   │       ├── stamp-h1
│   │   │       └── uncrustify_version.h
│   │   └── external_uncrustify-stamp
│   │       ├── external_uncrustify-build
│   │       ├── external_uncrustify-configure
│   │       ├── external_uncrustify-download
│   │       ├── external_uncrustify-mkdir
│   │       ├── external_uncrustify-patch
│   │       └── external_uncrustify-update
│   └── tmp
│       ├── external_uncrustify-cfgcmd.txt
│       └── external_uncrustify-cfgcmd.txt.in
└── Makefile

14 directories, 53 files
```

@dirk-thomas can you compare it to yours? If I find the `Makefile` which is having the issue and run `make` in the folder it is in, then I get the error because the `build/uncrustify/external_uncrustify-prefix/src/external_uncrustify-build` does not have the relatively specified `../make_token_names.sh` file.

**Comment by dirk-thomas** [comment]
The tree of my local build space looks like this (after having called `build` / `test` multiple times):

```
$ tree build/uncrustify
build/uncrustify
├── ament_cmake_args.cache
├── ament_cmake__build.sh
├── ament_cmake__install.sh
├── ament_cmake__test.sh
├── bin
│   └── uncrustify
├── CMakeCache.txt
├── CMakeFiles
│   ├── 2.8.12.2
│   │   ├── CMakeCCompiler.cmake
│   │   ├── CMakeCXXCompiler.cmake
│   │   ├── CMakeDetermineCompilerABI_C.bin
│   │   ├── CMakeDetermineCompilerABI_CXX.bin
│   │   ├── CMakeSystem.cmake
│   │   ├── CompilerIdC
│   │   │   ├── a.out
│   │   │   └── CMakeCCompilerId.c
│   │   └── CompilerIdCXX
│   │       ├── a.out
│   │       └── CMakeCXXCompilerId.cpp
│   ├── cmake.check_cache
│   ├── CMakeDirectoryInformation.cmake
│   ├── CMakeOutput.log
│   ├── CMakeRuleHashes.txt
│   ├── CMakeTmp
│   ├── external_uncrustify-complete
│   ├── external_uncrustify.dir
│   │   ├── build.make
│   │   ├── cmake_clean.cmake
│   │   ├── DependInfo.cmake
│   │   ├── depend.internal
│   │   ├── depend.make
│   │   └── progress.make
│   ├── Makefile2
│   ├── Makefile.cmake
│   ├── progress.marks
│   └── TargetDirectories.txt
├── cmake_install.cmake
├── external_uncrustify-prefix
│   ├── src
│   │   ├── external_uncrustify-build
│   │   │   ├── config.log
│   │   │   ├── config.status
│   │   │   ├── Makefile
│   │   │   ├── man
│   │   │   │   ├── Makefile
│   │   │   │   └── uncrustify.1
│   │   │   └── src
│   │   │       ├── config.h
│   │   │       ├── Makefile
│   │   │       ├── stamp-h1
│   │   │       ├── uncrustify
│   │   │       ├── uncrustify-align.o
│   │   │       ├── uncrustify-align_stack.o
│   │   │       ├── uncrustify-args.o
│   │   │       ├── uncrustify-backup.o
│   │   │       ├── uncrustify-brace_cleanup.o
│   │   │       ├── uncrustify-braces.o
│   │   │       ├── uncrustify-chunk_list.o
│   │   │       ├── uncrustify-ChunkStack.o
│   │   │       ├── uncrustify-combine.o
│   │   │       ├── uncrustify-compat_posix.o
│   │   │       ├── uncrustify-compat_win32.o
│   │   │       ├── uncrustify-defines.o
│   │   │       ├── uncrustify-detect.o
│   │   │       ├── uncrustify-indent.o
│   │   │       ├── uncrustify-keywords.o
│   │   │       ├── uncrustify-lang_pawn.o
│   │   │       ├── uncrustify-logger.o
│   │   │       ├── uncrustify-logmask.o
│   │   │       ├── uncrustify-md5.o
│   │   │       ├── uncrustify-newlines.o
│   │   │       ├── uncrustify-options.o
│   │   │       ├── uncrustify-output.o
│   │   │       ├── uncrustify-parens.o
│   │   │       ├── uncrustify-parse_frame.o
│   │   │       ├── uncrustify-punctuators.o
│   │   │       ├── uncrustify-semicolons.o
│   │   │       ├── uncrustify-sorting.o
│   │   │       ├── uncrustify-space.o
│   │   │       ├── uncrustify-tokenize_cleanup.o
│   │   │       ├── uncrustify-tokenize.o
│   │   │       ├── uncrustify-uncrustify.o
│   │   │       ├── uncrustify-unc_text.o
│   │   │       ├── uncrustify-unicode.o
│   │   │       ├── uncrustify-universalindentgui.o
│   │   │       ├── uncrustify_version.h
│   │   │       └── uncrustify-width.o
│   │   └── external_uncrustify-stamp
│   │       ├── external_uncrustify-build
│   │       ├── external_uncrustify-configure
│   │       ├── external_uncrustify-done
│   │       ├── external_uncrustify-download
│   │       ├── external_uncrustify-install
│   │       ├── external_uncrustify-mkdir
│   │       ├── external_uncrustify-patch
│   │       └── external_uncrustify-update
│   └── tmp
│       ├── external_uncrustify-cfgcmd.txt
│       └── external_uncrustify-cfgcmd.txt.in
├── install_manifest.txt
├── Makefile
└── share
    ├── man
    │   └── man1
    │       └── uncrustify.1
    └── uncrustify
        ├── ben.cfg
        ├── d.cfg
        ├── defaults.cfg
        ├── gnu-indent.cfg
        ├── kr-indent.cfg
        ├── linux.cfg
        ├── linux-indent.cfg
        ├── mono.cfg
        └── xsupplicant.cfg

18 directories, 98 files
```

**Comment by dirk-thomas** [comment]
The missing file (`make_token_names.sh`) is in the source space - a sibling of the package.xml file: https://github.com/ament/uncrustify/blob/b9cd1dc9b73eab61904491bf2a565622b34e5cc4/make_token_names.sh

**Comment by wjwwood** [comment]
The issue is that my `src` folder is a symlink. I'm not sure how we can work around this, but I think it is pretty clearly a CMake issue. I'll try a few things to see if I can coerce it into working.

**Comment by dirk-thomas** [comment]
I had the same problem again on a fresh build (actually twice). It looks like that there is still a race condition. Therefore reopening it.

**Comment by dirk-thomas** [comment]
Addressed by #5.
